core/src/main/scala/producer/KafkaProducer.scala

```diff
@@ -0,0 +1,186 @@
+
+object producer {
+
+  final class KafkaProducerProvider[K, V] {
+
+    @free
+    trait Producer {
+
+      def isClosed: FS[Boolean]
+
+      def close(): FS[Unit]
+
+      def closeWaitingFor(timeout: Duration): FS[Unit]
+
+      def flush(): FS[Unit]
+
+      def metrics: FS[Map[MetricName, Metric]]
+
+      def partitionsFor(topic: Topic): FS[List[PartitionInfo]]
+
+      def send(record: ProducerRecord[K, V]): FS[RecordMetadata]
+
+      def sendToTopic(topic: Topic, record: (K, V)): FS[RecordMetadata]
+
+      def sendMany(records: List[ProducerRecord[K, V]]): FS[List[RecordMetadata]]
+
+      def sendManyToTopic(topic: Topic, records: List[(K, V)]): FS[List[RecordMetadata]]
+
+      def initTransactions(): FS[Unit]
+
+      def beginTransaction(): FS[Unit]
+
+      def sendOffsetsToTransaction(
+          offsets: Map[TopicPartition, OffsetAndMetadata],
+          consumerGroupId: String): FS[Unit]
+
+      def commitTransaction(): FS[Unit]
+
+      def abortTransaction(): FS[Unit]
+
+    }
+
+    class KafkaProducerHandler[M[_]](
+        implicit underlying: UnderlyingKafkaProducer[K, V],
+        ctx: AsyncContext[M],
+        ME: MonadError[M, Throwable])
+        extends Producer.Handler[M] {
+
+      private val atomicProducer: AtomicReference[KafkaProducer[K, V]] =
+        new AtomicReference(underlying.producer)
+
+      private val producerClosedState: AtomicReference[Boolean] =
+        new AtomicReference(false)
+
+      def producer: KafkaProducer[K, V] = {
+        if (producerClosedState.get()) {
+          atomicProducer.set(underlying.producer)
+          producerClosedState.set(false)
+        }
+        atomicProducer.get()
+      }
+
+      override protected[this] def isClosed: M[Boolean] =
+        producerClosedState.get().pure[M]
+
+      override protected[this] def close: M[Unit] = {
+        ME.catchNonFatal(producer.close()).map { _ =>
+          producerClosedState.set(true)
+          ()
+        }
+      }
+
+      override protected[this] def closeWaitingFor(timeout: Duration): M[Unit] = {
+        ME.catchNonFatal(producer.close(timeout.toNanos, TimeUnit.NANOSECONDS)).map { _ =>
+          producerClosedState.set(true)
+          ()
+        }
+      }
+
+      override protected[this] def flush: M[Unit] = ME.catchNonFatal(producer.flush())
+
+      override protected[this] def metrics: M[Map[MetricName, Metric]] =
+        ME.catchNonFatal(producer.metrics().asScala.toMap[MetricName, Metric])
+
+      override protected[this] def partitionsFor(topic: Topic): M[List[PartitionInfo]] =
+        ME.catchNonFatal(producer.partitionsFor(topic).asScala.toList)
+
+      override protected[this] def send(record: ProducerRecord[K, V]): M[RecordMetadata] = {
+        ctx.runAsync { cb =>
+          producer
+            .send(
+              record,
+              new Callback {
+                override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+                  println(
```

Comment:
Given we've set up a logback file in the testing scope, can we replace this println statement by a proper log statement.